### PR TITLE
Fixed backface visibility.

### LINKED ISSUE #2985: backface-visibility: hidden doesn't hide elements from cypress
### Current behavior:

`cy.contains()` is getting an element which isn't actually visible

### Desired behavior:

It should fail as the element isn't visible

### Steps to reproduce: (app code and test code)

https://codepen.io/desandro/pen/LmWoWe

On this pen, `cy.contains('back')` will pass even when it is hidden.

### Versions

Latest version, chrome

### PR DESCRIPTION
**Pull request by sainthkh**
- Closes #2985

### User facing changelog

When `backface-visibility: hidden` and `rotateX(180deg)`, `to.be.hidden` is now `true`. 

### Additional details

- Why was this change necessary?

[Like card example provided](https://codepen.io/desandro/pen/LmWoWe), sometimes we want something to be invisible when it is rotateX(180deg) or rotateY(180deg). 

Now, we can check them out. 

- What is affected by this change?

I just added some tests and related code. 

- Any implementation details to explain?

It uses simplified version of [backface culling](https://en.wikipedia.org/wiki/Back-face_culling). 

### How has the user experience changed?

Now we can check backface. 

### PR Tasks
- [x] Have tests been added/updated?
- ~~Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? ~~
- ~~Have API changes been updated in the [`type definitions`](cli/types/index.d.ts)?~~
- ~~Have new configuration options been added to the [`cypress.schema.json`](cli/schema/cypress.schema.json)?~~

### CODE REVIEW

**Comment by sainthkh** [comment]
retest

**Comment by sainthkh** [comment]
I guess `big cookie` test failed because it's flaky. 

But interestingly, it doesn't work on my local ubuntu with develop branch code.

**Comment by Bkucera** [comment]
@sainthkh what happens if the element has `rotateY(180deg)` and its parent has `rotateY(-180deg)`? I believe the transform property will say 180 degrees when in reality the element would be visible.

**Comment by sainthkh** [comment]
@Bkucera I'll check it out.

**Comment by sainthkh** [comment]
@Bkucera Interestingly, that didn't happen. 

I created [this code pen](https://codepen.io/sainthkh/pen/WNNMagM) and tested with Chrome 78 and Firefox 70 on Ubuntu 18.04. I couldn't see "hello transformed world". 

p.s. server-integration-tests failed because of timeout.

**Comment by Bkucera** [comment]
thanks @sainthkh , I'll rerun CI

File: packages/driver/src/dom/visibility.js [hunk 1]
```diff
@@ -112,6 +116,42 @@ const elHasVisibilityHidden = ($el) => {
...
+// We defined view normal vector, (0, 0, -1), - eye to screen.
+// and default normal vector, (0, 0, 1)
+// When dot product of them are >= 0, item is visible.
+const elIsBackface = ($el) => {
+  const el = $el[0]
+  const style = getComputedStyle(el)
+  const backface = style.getPropertyValue('backface-visibility')
+  const backfaceInvisible = backface === 'hidden'
+
+  if (backfaceInvisible) {
```
Comment (chrisbreiding) [inline packages/driver/src/dom/visibility.js]: Let's reduce the nesting here by flipping the logic and returning early. 

```javascript
// ...
const backfaceInvisible = backface === 'hidden' 
const transform = style.getPropertyValue('transform')

if (!backfaceInvisible || !transform.startsWith('matrix3d')) {
  return false
}

// ... rest can be un-nested ...
```

File: packages/driver/src/dom/visibility.js [hunk 2]
```diff
@@ -112,6 +116,42 @@ const elHasVisibilityHidden = ($el) => {
...
+    const transform = style.getPropertyValue('transform')
+
+    if (transform.startsWith('matrix3d')) {
+      const m3d = transform.substring(8).match(/-?[0-9]+(?:\.[0-9]+)?/g)
```
Comment (chrisbreiding) [inline packages/driver/src/dom/visibility.js]: Pull this regex out to the root level into a `const` so it doesn't get re-instantiated every time this function is run and so it's given a name.

**Comment by sainthkh** [comment]
@chrisbreiding Fixed.

**Pull request merged**